### ISSUE TITLE
Use `fetch` from Open Terms Archive Core in Contribution UI

### ISSUE THREAD

**Issue by MattiSG** [OpenTermsArchive/contribution-tool#39]
- Expose `fetch` as a module in Open Terms Archive Core.
- Use `fetch` from Open Terms Archive Core to obtain the content in the Contribution UI.

Comments from Martin
-------
This is in fact a bit more complicated than it appears
I am using several techniques that are not present on the core today, and especially
- Recaptcha plugin
- use of the code within chrome extension "i-dont-care-about-cookies" to remove the cookie banners from the page
- rewrite of urls in the scraped html code to be able to display it correctly in an iframe (to bypass CORS)

So I would suggest to leave this for later for now 

Comments after discussion
-------------------

The main point is to ensure alignment between what is displayed to the user and what is fetched by Open Terms Archive, so that the contributed declarations do not systematically rely on `executeClientScripts`.

**Comment by martinratinaud** [comment]
For the record, I have faced many problems

Here is the PR for exposing features from OTA-Core https://github.com/ambanum/OpenTermsArchive/pull/852

## import ES module in NextJs custom server

I created an issue with reproduceable steps on NextJs https://github.com/vercel/next.js/issues/36940
TLDR: a module exposing features using the package.json `exports` feature is unusable with nextjs custom server

## typescript on server

In order to be able to use a custom server, tsconfig must be overcharged using `"module": "commonjs",` because express was built on the commonjs pattern (See https://levelup.gitconnected.com/set-up-next-js-with-a-custom-express-server-typescript-9096d819da1c)
I tried with fastify ans koa to see if it would work without this trick, but I ended up with the same problem.
This makes it impossible (at least for me) to import a ES module as the one from open-terms-archive

## import WITHOUT exports functionality

I tried to remove the `exports` from `package.json` and include directly the file in the project like
```
import fetcher from 'open-terms-archive/src/archivist/fetcher/index.js';
```
but it fails with 
```
../OpenTermsArchive/node_modules/clone-deep/utils.js
Cannot statically analyse 'require(…, …)' in line 11
```

So I'm a bit stuck and need to pair program with one of you to understand what's going on.


 

https://github.com/vercel/next.js/issues/36940

**Comment by martinratinaud** [comment]
This is thus soon to be merged on the Core side with https://github.com/ambanum/OpenTermsArchive/pull/867

On the UI side, (for information, not ready to be merged), here is a first imple
mentation (and WIP) which needs UI by @clementbiron on [`ota_fetch`](https://github.com/OpenTermsArchive/contribution-tool/tree/ota_fetch)

I added a "verify" button that opens side by side the snapshot and version.


![version-ota](https://user-images.githubusercontent.com/4191809/170484490-609c9808-1d89-4a98-984e-a04ae6b0d085.gif)

@clementbiron, please, let's talk based on this first shot.

**Comment by martinratinaud** [comment]
After days of research is what I ended up with

## Docker
Deployment was made through docker on the OTA main server, and was using an arg from puppeteer called "--no-sandbox".
This seems to be a bad practice https://no-sandbox.io/ but still contained as the docker container is not executed as Root but as a custom user with very few privileges.

I crawled the web in search for a solution to this specific problem and found many which were all not satisfying:
- Tried all this: https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md with no success
- launch docker with `-privileged`
- using a [custom seccomp](https://stackoverflow.com/a/53975412/2497753) that removes some security checks

After discussion with @Ndpnt, keeping `--no-sandbox` seems not recommended (even though it is the case right now).

## Scalingo

I thus tried to deploy on scalingo, which would be a lot better with less maintenance.
Applied this documentation https://doc.scalingo.com/languages/nodejs/puppeteer and ended up with the same problem.
After talking to scalingo support, it appears it is also needed to pass `--no-sandbox` (They will update the documentation soon)

In order to prove it can work, I finally had a working version on Scalingo
https://opentermsarchive-contribution-tool.osc-fr1.scalingo.io/en/service?destination=OpenTermsArchive/sandbox&localPath=/Users/martin/Workspace/ambanum/contrib-declarations/declarations&versionsRepo=OpenTermsArchive/contrib-versions&url=https%3A%2F%2Fwww.bizzbee.com%2Fcgv.html

passing `--no-sandbox` as an environment variable

I can't help wondering how Scalingo can allow such a thing if it is that dangerous though. I would be happy to have your thoughts on this.

## Ansible

In order for the code to be able to be deployed on our server, I removed the docker based deployment and added a direct ansible deployment.
FI, on the OTA server, there is now
- NGINX listening to *.opentermsarchive.org dispatching the following items
- preprod contribution-tool deployed as a node app on the server
- prod contribution-tool deployed as docker (but will be another node app once this PR is merged)
- OpenTermsArchive.org as docker which will be transfered to scalingo
- Preprod.opentermsarchive.org as docker which will be discontinued
- OpenTermsArchive legacy API

I still work on the PR then with this in mind

**Comment by martinratinaud** [comment]
Here is the answer from scalingo in french 

> En effet, si vous utilisez des sources non fiables, il y a un risque de sécurité pour votre application (un site web avec un code malveillant pourrait techniquement exécuter du code dans le conteneur).
Sur des environnements comme Scalingo, c'est la seule façon de faire fonctionner Puppeteer. Il serait possible de notre côté d'être compatible avec la sandbox Chrome, mais cela demande des changements majeurs de notre côté et je ne pourrais pas vous dire si cela sera fait et encore moins dans quelle durée.
Je vais mettre à jour notre documentation pour refléter cela.

So I guess this settles it and we're keeping it on our own server.

**Comment by martinratinaud** [comment]
2 PRs are finally awaiting approval

- one with refactors and fixes https://github.com/OpenTermsArchive/contribution-tool/pull/45
- one with the actual code https://github.com/OpenTermsArchive/contribution-tool/pull/44

**Comment by MattiSG** [comment]
> I can't help wondering how Scalingo can allow such a thing if it is that dangerous though. I would be happy to have your thoughts on this.

Because they are not responsible for what you do to the server they make available to you, no? What do they care if our app is insecure. This is our problem 😉

If we are worried about the contribution tool security, why not host it on its own server and use a CNAME instead of NGINX to serve the dedicated subdomain? 🙂
